homeassistant/components/balboa/fan.py

```diff
@@ -0,0 +1,102 @@
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.percentage import (
+    percentage_to_ranged_value,
+    ranged_value_to_percentage,
+)
+
+from .const import DOMAIN
+from .entity import BalboaEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the spa's pumps."""
+    spa: SpaClient = hass.data[DOMAIN][entry.entry_id]
+    entities = [BalboaPumpFanEntity(control) for control in spa.pumps]
+    async_add_entities(entities)
+
+
+PUMP_ICON_ON = "mdi:pump"
+PUMP_ICON_OFF = "mdi:pump-off"
+
+
+class BalboaPumpFanEntity(BalboaEntity, FanEntity):
+    """Representation of a Balboa Spa pump fan entity."""
+
+    _attr_supported_features = FanEntityFeature.SET_SPEED
+
+    def __init__(self, control: SpaControl) -> None:
+        """Initialize a Balboa pump fan entity."""
+        super().__init__(control.client, control.name)
+        self._control = control
+        if len(control.client.pumps) > 1:
+            self._attr_translation_key = "pump_of_n"
+            self._attr_translation_placeholders = {
+                "index": f"{cast(int, control.index) + 1}"
+            }
+        else:
+            self._attr_translation_key = "only_pump"
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the pump off."""
+        await self._control.set_state(OffOnState.OFF)
+
+    async def async_turn_on(
+        self,
+        percentage: int | None = None,
+        preset_mode: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn the pump on (by default on max speed)."""
+        if percentage is None:
+            percentage = 100
+        await self.async_set_percentage(percentage)
+
+    async def async_set_percentage(self, percentage: int) -> None:
+        """Set the speed of the pump."""
+        if percentage > 0:
+            state = math.ceil(
+                percentage_to_ranged_value((1, self.speed_count), percentage)
+            )
+        else:
+            state = OffOnState.OFF
+        await self._control.set_state(state)
+
+    @property
+    def percentage(self) -> int | None:
+        """Return the speed of the pump."""
+        if self._control.state == UnknownState.UNKNOWN:
+            return None
+        if self._control.state == OffOnState.OFF:
+            return 0
+        return ranged_value_to_percentage((1, self.speed_count), self._control.state)
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return true if the pump is running."""
+        if self._control.state == UnknownState.UNKNOWN:
+            return None
+        return self._control.state != OffOnState.OFF
+
+    @property
+    def icon(self) -> str | None:
+        """Return an icon representing a pump, repectively in the on or off position."""
+        return PUMP_ICON_ON if self.is_on else PUMP_ICON_OFF
```

Comment:
Please use icon translations: https://developers.home-assistant.io/docs/core/entity#icons

Reply:
How does that work, now where I use two different translation keys based on whether it's the only pump or one of multiple? Do I have to duplicate the icon translation for each translation key?

(FWIW, I just did this consistent with the existing code in this integration which already uses the icon property in `binary_sensor.py` and `climate.py`.)

Reply:
> How does that work, now where I use two different translation keys based on whether it's the only pump or one of multiple? Do I have to duplicate the icon translation for each translation key?

Yes, you need to duplicate the icon translation for each translation key.

> (FWIW, I just did this consistent with the existing code in this integration which already uses the icon property in binary_sensor.py and climate.py.)

It's true, we haven't migrated every integration yet. The icon translations are quite new (introduced with 2024.2)

Reply:
Done. I got rid of the duplication, not sure if there is a model with only one pump anyway.